defaults.go

```diff
@@ -6,17 +6,23 @@ import (
 	"errors"
 	"reflect"
 	"strconv"
+	"strings"
 	"time"
 )
 
 var (
+	errInvalidTag  = errors.New("invalid tag name")
 	errInvalidType = errors.New("not a struct pointer")
 )
 
 const (
 	fieldName = "default"
 )
 
+var (
+	TagName string = fieldName
+)
```

Comment:
Why adding a new exported variable

What about this ?

```suggestion
 var (
 	fieldName = "default"
 )
 
func SetFieldName(name string) {
   fieldName = name
}
```

Reply:
Idiomatically, I've only seen it done the way provided - a public variable with an initialized default of an unexported constant.

I'm not sure why a function would be beneficial over a variable, though I suppose validation of the tag name could be moved there. I just typically don't see it done that way.

Reply:
My problem is the possible race conditions, and bug for people setting the default value at package level.

Also, imagine 2 libs using Set after passing the TagName

The more I'm thinking about it, the more I would remove this variable and only add SetWithTag

Think about the test in the lib, you cannot run parallel tests when testing TagName

Reply:
> My problem is the possible race conditions, and bug for people setting the default value at package level.

Unless a mutex is implemented, which isn't present in your alternative, `defaults.TagName = "foo"` is susceptible to the exact same race conditions as `defaults.SetTagName("foo")`, no?

> Also, imagine 2 libs using Set after passing the TagName

Sure, valid. Again, this is the idiomatic way I've seen tag name overrides; I'm not saying it's the *best* (we both know idiomatic frequently isn't the best).

> The more I'm thinking about it, the more I would remove this variable and only add SetWithTag

A fair evaluation, and probably the safest. I'd assume it'd be sensible then to change `Set()` to wrap `SetWithTag()` using the package constant `default` name then to reduce maintenance cost?

Reply:
> > My problem is the possible race conditions, and bug for people setting the default value at package level.
> 
> Unless a mutex is implemented, which isn't present in your alternative, `defaults.TagName = "foo"` is susceptible to the exact same race conditions as `defaults.SetTagName("foo")`, no?

It was my next move 😁

> > Also, imagine 2 libs using Set after passing the TagName
> 
> Sure, valid. Again, this is the idiomatic way I've seen tag name overrides; I'm not saying it's the *best* (we both know idiomatic frequently isn't the best).
> 
> > The more I'm thinking about it, the more I would remove this variable and only add SetWithTag
> 
> A fair evaluation, and probably the safest. I'd assume it'd be sensible then to change `Set()` to wrap `SetWithTag()` using the package constant `default` name then to reduce maintenance cost?

definitely

Reply:
FIXED:
* No use of global consts/vars anymore, nor a function to set such a global var
* Done entirely via SetWithTag and MustSetWithTag
* MustSet and MustSetWithTag wrap the above respectively